### PR TITLE
Implement SlidingWindowReadWalker for sliding-window analysis

### PR DESCRIPTION
**Pull request by magicDGS**
Includes:

* `ShardingIterator` to avoid too many query calls
* `ReadSliderWalker` with documentation on how to implement (edited: walker name is `SlidingWindowReadWalker` instead)
* Argument collection interface for sharding arguments
* Example walker and argument collection
* Tests for example walker

Closes https://github.com/broadinstitute/gatk/issues/1198

### CODE REVIEW

**Comment by magicDGS** [comment]
@droazen - finally I have the implementation for the long-waiting sliding-window walker (after several failed trials). This one does not have any interference with the classes in any tool from GATK to speed-up acceptance.

Could you give me a rough estimate of the time for acceptance of this?

**Comment by droazen** [comment]
@magicDGS Given that it's much more self-contained this time around, I think it should not take too long to get in. We'll try to get you a review within the next few days.

**Comment by magicDGS** [comment]
Thanks @droazen!

**Comment by magicDGS** [comment]
@droazen - sorry for the compilation error, it was just an early optimization. Can you have a look to it? Thanks!

**Comment by magicDGS** [comment]
Finally got the test passing in travis (silly error from refactoring previous stuff). Ready to go, @droazen!

**Comment by magicDGS** [comment]
Friendly ping here @droazen!

**Comment by magicDGS** [comment]
@droazen - any idea about the time to get into this? It has been 2 years since the original proposal (https://github.com/broadinstitute/gatk/pull/1528) and almost a month for this to be reviewed. I understand that the bandwidth for community-driven development is lower than the one for broad needs - but I need to get non-intrusive functionality like this in an easier way to be shared between downstream projects.

I'll probably opt to create a new project for this kind of gatk extensions: general-purpose walkers, common utility for gatk abstractions (e.g., `GATKRead`), etc. In that case, I am open to backport anything that might be useful for your team (and for the rest of the community), but I cannot block my own development due to long-standing PRs.

Let me know what do you think about this. I will keep contributing to GATK and proposing new code to be included, but I guess that having a repository with this code available for the community in a versioned and delivered way might be useful.

**Comment by droazen** [comment]
Sorry @magicDGS -- this slid off my radar. I'll get you a review today.

File: src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+                nextData.isEmpty() ? Collections.emptyList() : new LinkedList<>(nextData));
+        advance();
+        return next;
+    }
+
+    // advance to the next data
+    private void advance() {
+        // if we already finished the iteration, we clean up the data
+        if (shards.isEmpty()) {
+            // TODO: should we also close the iterator?
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java]: Well, your iterator doesn't take a `CloseableIterator`, so you can't close it. Our iterator classes don't typically close their nested iterators on traversal end anyway -- it's the client's responsibility to call `close()` at traversal end when dealing with `CloseableIterators`, so don't worry about it here.

File: src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            // then, fill in the data
+            fillQueueWithOverlapping();
+        }
+    }
+
+    // assumes that shards are not empty
+    private void removeNonOverlapping() {
+        while (!nextData.isEmpty() && !nextData.peek()
+                .overlaps(shards.peek().getPaddedInterval())) {
+            nextData.poll();
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java]: This is not quite right. The reads could be of varying lengths (because of clipping, for example), and so you can't stop as soon as you encounter the *first* read that overlaps the next shard's interval, because there could be a *shorter* read after it that doesn't overlap the next interval.

You need to check *all* reads in `nextData` that start before the next shard's padded interval, and remove those that don't overlap with it. You can stop only once you encounter a read whose start position is >= the start of the next shard's padded interval. See `FeatureCache.trimToNewStartPosition()` for a similar algorithm.

Once you've made this change, you should add a unit test that covers the edge case described above.

File: src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        }
+    }
+
+    // assumes that shards are not empty
+    private void fillQueueWithOverlapping() {
+        while (it.hasNext()) {
+            // if this one overlaps
+            if (it.peek().overlaps(shards.peek().getPaddedInterval())) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java]: Your iterator will not work with unmapped reads, which implies that you should have the `MappedReadFilter` on by default in your walker base class (in addition to the `WellformedReadFilter`, which is already on).

File: src/test/java/org/broadinstitute/hellbender/utils/iterators/ShardingIteratorUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+        // create shard iterator
+        // create a list of reads larger than the interval
+        final ShardingIterator<GATKRead> it = new ShardingIterator<>(
+                reads.iterator(),
+                shards,
+                TEST_HEADER.getSequenceDictionary());
+
+        for (final ShardBoundary s: shards) {
+            testNextElement(it, s.getInterval(), s.getPaddedInterval(), s.getPaddedInterval().getLengthOnReference());
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/iterators/ShardingIteratorUnitTest.java]: Also assert that you get the correct number of shards back from your iterator (applies to your other tests as well).

File: src/test/java/org/broadinstitute/hellbender/utils/iterators/ShardingIteratorUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        // create shard iterator
+        // create a list of reads larger than the interval
+        final ShardingIterator<GATKRead> it = new ShardingIterator<>(
+                reads.iterator(),
+                shards,
+                TEST_HEADER.getSequenceDictionary());
+
+
+        for (final ShardBoundary s: shards) {
+            testNextElement(it, s.getInterval(), s.getPaddedInterval(), s.getPaddedInterval().getLengthOnReference());
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/iterators/ShardingIteratorUnitTest.java]: Assert that you get the correct number of shards back.

File: src/test/java/org/broadinstitute/hellbender/utils/iterators/ShardingIteratorUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+import org.testng.annotations.Test;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class ShardingIteratorUnitTest extends GATKBaseTest {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/iterators/ShardingIteratorUnitTest.java]: This iterator needs more/better tests, in addition to the tests you have below. In particular, I think you need to:

* Test with reads of different lengths, sorted by start position. As noted above, I think you have a bug in your iterator in this case.

* Test with realistically-sized (~75 bp or longer) reads that overlap, instead of just 1-base non-overlapping reads.

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/OptionalShardingArgumentCollection.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
+package org.broadinstitute.hellbender.cmdline.argumentcollections;
+
+import org.broadinstitute.barclay.argparser.Argument;
+import org.broadinstitute.barclay.argparser.CommandLineException;
+
+/**
+ * Implementation of {@link ShardingArgumentCollection} with optional values for all arguments.
+ *
+ * <p>Note: this class use the recommended argument names defined in {@link ShardingArgumentCollection}.
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/OptionalShardingArgumentCollection.java]: Done

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/ShardingArgumentCollection.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    /** Recommended valur for window-pad argument. */
+    public static final String WINDOW_PAD_NAME = "window-pad";
+
+    /**
+     * Returns the window-size applied to the data.
+     */
+    public int getWindowSize();
+
+    /**
+     * Returns the window-step applied to the data.
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/ShardingArgumentCollection.java]: Done

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/ShardingArgumentCollection.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+     *
+     * <p>If equal to {@link #getWindowSize()}, the analysis would be performed in non-overlapping windows.
+     */
+    public int getWindowStep();
+
+    /**
+     * Returns the window-padding applied to the data.
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/ShardingArgumentCollection.java]: Done

File: src/main/java/org/broadinstitute/hellbender/engine/ReadSliderWalker.java [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+ * <p>{@link ReadSliderWalker} authors must implement the {@link #apply(Shard, ReferenceContext, FeatureContext)}
+ * and {@link #getShardingArgs()} methods, and optionally implement {@link #onTraversalStart()}
+ * and/or {@link #onTraversalSuccess()}. For closing other sources, such as output writers,
+ * {@link #closeTool()} might be implemented.
+ *
+ * <p>For an example walker, see {@link org.broadinstitute.hellbender.tools.examples.ExampleReadSliderWalker}.
+ *
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public abstract class ReadSliderWalker extends GATKTool {
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/engine/ReadSliderWalker.java]: Done (also changed the title of the PR)

File: src/main/java/org/broadinstitute/hellbender/engine/ReadSliderWalker.java [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A {@link ReadSliderWalker} is a tool that processes a single {@link Shard<GATKRead>} window over
+ * the genome/intervals at a time, with the ability to query overlapping optional sources (reference
+ * data and/or variants/features).
+ *
+ * <p>Windows are constructed over each interval, using the window-size, window-step and window-padding
+ * provided by the {@link #getShardingArgs()} arguments.
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/engine/ReadSliderWalker.java]: Done. But due to this comment, I am thinking about a different design for client flexibility: delegate window-construction to the argument collection (through the method `makeWindows(List<SimpleInterval>, SAMSequenceDictionary)`) and convert the `SlidingWindowReadWalker` into a `ReadShardWalker`. This have the following advantages for the client:

* Flexibility designing how to construct the shards. For example, I can imagine a tool which iterates over pre-defined windows loaded from a feature file specified by the user and certain padding.
* This removes the description on the window size/step/padding parameter, which are quite specific to my use case. The client might choose which arguments are exposed in an easier way.

For my current use case, this walker is flexible enough. Let me know if by your end you have any feeling about changing the implementation to support this flexibility. Maybe I am overthinking about possible usages and that has not usability at all.

File: src/main/java/org/broadinstitute/hellbender/engine/ReadSliderWalker.java [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    protected final void onStartup() {
+        super.onStartup();
+
+        // validate first the sharding arguments
+        shardingArgs.validate();
+
+        // get the dictionary and check that it is provided
+        final SAMSequenceDictionary dictionary = getBestAvailableSequenceDictionary();
+        if (dictionary == null) {
+            throw new UserException(String.format("Could not find sequence dictionary in data sources.A dictionary file is necessary for {}",
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/engine/ReadSliderWalker.java]: Done. Oops - I always make this kind of errors when writing messages using both `String.format` and a logger.

File: src/main/java/org/broadinstitute/hellbender/engine/ReadSliderWalker.java [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+                    this.getToolName()));
+        }
+
+        // set traversal bounds if necessary
+        if (hasIntervals() && hasReads()) {
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/engine/ReadSliderWalker.java]: This comes from a previous implementation where it was a more general walker over shards (reads/reference/features without any of them required). Removed!

File: src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleReadSliderWalker.java [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+        return new OptionalShardingArgumentCollection(DEFAULT_WINDOW_SIZE, DEFAULT_WINDOW_STEP, DEFAULT_WINDOW_PAD);
+    }
+
+    @Override
+    public void apply(final Shard<GATKRead> reads, final ReferenceContext reference, final FeatureContext features) {
+        // compute read coverage over window
+        final long readCoverage = Utils.stream(reads.iterator()).count();
+
+        // output coverage  over window
+        outputStream.printf("%s (%s padded): %s reads.",
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleReadSliderWalker.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+ *
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public final class ShardingIterator<T extends Locatable> implements Iterator<Shard<T>> {
+
+    // underlying iterator - should be sorted by coordinate
+    private final PeekableIterator<T> it;
+    // queuing the shards - should be sorted by coordinate
+    private final Queue<ShardBoundary> shards;
+    // this wil be fill with the data for the shard in the first position of the shard queue
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+            final SAMSequenceDictionary dictionary) {
+        // check arguments
+        Utils.validateArg(Utils.nonNull(iterator, "null iterator").hasNext(), "empty iterator");
+        Utils.nonEmpty(boundaries, "Empty shard boundaries");
+        Utils.nonNull(dictionary, "null dictionary");
+
+        // TODO: avoid new object creation for already PeekableIterator and LinkedList args
+        // store the iterator and allow peek()
+        this.it = new PeekableIterator<>(iterator);
+        // TODO: avoid new object creation for already LinkedList and LinkedList args
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java]: Removed.

File: src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+
+/**
+ * Iterator over shards/windows (overlapping or not) of sorted {@link Locatable}.
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java (hunk 1 above)
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java]: I was thinking about using `htsjdk.samtools.util.CloserUtil`, but I guess that it is better to do not do it in this case. Removed TODO.

File: src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+     *
+     * @param iterator   sorted iterator to iterate over windows.
+     * @param boundaries sorted list of shard boundaries to iterate over.
+     * @param dictionary dictionary for position comparison.
+     */
+    public ShardingIterator(final Iterator<T> iterator,
+            final List<ShardBoundary> boundaries,
+            final SAMSequenceDictionary dictionary) {
+        // check arguments
+        Utils.validateArg(Utils.nonNull(iterator, "null iterator").hasNext(), "empty iterator");
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java]: That's the right way to go, I agree. Removed the check for `.hasNext`

File: src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+        Utils.nonEmpty(boundaries, "Empty shard boundaries");
+        Utils.nonNull(dictionary, "null dictionary");
+
+        // TODO: avoid new object creation for already PeekableIterator and LinkedList args
+        // store the iterator and allow peek()
+        this.it = new PeekableIterator<>(iterator);
+        // TODO: avoid new object creation for already LinkedList and LinkedList args
+        // convert the boundaries into a queue
+        this.shards = new LinkedList<>(boundaries);
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java]: Changed to `ArrayDeques` for both.

File: src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+     * next iteration.
+     */
+    @Override
+    public Shard<T> next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+        // copying the deque to being able to remove unnecesary data and fill in with the next
+        final Shard<T> next = new ShardBoundaryShard<T>(shards.poll(),
+                nextData.isEmpty() ? Collections.emptyList() : new LinkedList<>(nextData));
```
[the remark this answers is not in the document]
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java (hunk 3 above)
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java]: I added already the `MappedReadFilter`. I might consider to add unmapped reads too, but not at the moment.

File: src/test/java/org/broadinstitute/hellbender/utils/iterators/ShardingIteratorUnitTest.java (hunk 4 above)
Comment (magicDGS) [reply src/test/java/org/broadinstitute/hellbender/utils/iterators/ShardingIteratorUnitTest.java]: I think that it is implicitly tested because I am iterating over the `ShardBoundary` list: if the number of elements returned is smaller it will fail because there is no element; if the number of elements is smaller, the next assertion of hasNext will fail.

Should I add explicitly a test for this by adding an `int` variable and assert that it is equal to the `shards.size()`?

File: src/test/java/org/broadinstitute/hellbender/utils/iterators/ShardingIteratorUnitTest.java (hunk 6 above)
Comment (magicDGS) [reply src/test/java/org/broadinstitute/hellbender/utils/iterators/ShardingIteratorUnitTest.java]: Does the realistically-sized test adds something to the unit test itself? I think that the itegration test with realistic BAM files is enough.

I will definitely implement the other test case before fixing the bug to demonstrate the issue and that the fix works.

File: src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java (hunk 2 above)
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java]: Good catch! I am working now on this, and I will let you know when it is prepared for next review.

Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java]: Done (failing test in one commit, and fix in the next)

**Comment by droazen** [comment]
@magicDGS Is this one ready for a second review pass?

**Comment by magicDGS** [comment]
Not yet - I didn't have the time to figure out how to do the test utility to create small pile of reads (because I think that there is already similar functionality in `ArtificialReadUtils`.

**Comment by droazen** [comment]
@magicDGS In the interest of moving this PR along, perhaps it would be easier to just write an integration test for `ExampleSlidingWindowReadWalker` that uses a large interval from a more realistic bam (such as `CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam` from `large`). This would address most of my concerns, and we could create a ticket to eventually add exhaustive tests with synthetic reads in a future PR.

**Comment by droazen** [comment]
Closing this PR as abandoned.
